Add support for converting a String to an Instant

What I'm trying to achieve:

```java
@ConfigurationProperties
record ApplicationProps(Instant instant) {}
```

with the following `application.properties`:
```
instant = 1680612880417
```
where the value is millis since the epoch, see [`Instant.ofEpochMilli()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/Instant.html#ofEpochMilli(long)).

Edit:
Currently this fails with a
```
...
Caused by: java.lang.IllegalArgumentException: Parse attempt failed for value [123456789]
	at org.springframework.format.support.FormattingConversionService$ParserConverter.convert(FormattingConversionService.java:224) ~[spring-context-6.0.6.jar:6.0.6]
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:41) ~[spring-core-6.0.6.jar:6.0.6]
	... 117 common frames omitted
Caused by: java.time.format.DateTimeParseException: Text '123456789' could not be parsed at index 0
```

Perhaps this is a good candidate for a default binding.

Considerations:
- Why are we choosing `Instant.ofEpochMilli()` and not `Instant.ofEpochSecond()`? It just felt more natural. Is this uncertainty reason enough to refuse this request? That said, there already is a registered-by-default `Converter<Long, Instant>` which uses `ofEpochMilli`. Therefore, an additional Formatter would be consistent with the already-existing Converter.
- Do we also need to add more similar formatters for types like `java.util.Date` or `java.sql.Timestamp`? I do no think think so as neither of them has a simple `ofEpochMilli()`-like creator method or clear semantics. Other `java.time` also are not good targets for this unless OffsetDateTime/ZonedDateTime is considered okay given a default timezone.

--- a/spring-context/src/main/java/org/springframework/format/datetime/standard/InstantFormatter.java
+++ b/spring-context/src/main/java/org/springframework/format/datetime/standard/InstantFormatter.java
@@ -41,13 +41,16 @@ public class InstantFormatter implements Formatter<Instant> {
 
 	@Override
 	public Instant parse(String text, Locale locale) throws ParseException {
-		if (text.length() > 0 && Character.isAlphabetic(text.charAt(0))) {
-			// assuming RFC-1123 value a la "Tue, 3 Jun 2008 11:05:30 GMT"
-			return Instant.from(DateTimeFormatter.RFC_1123_DATE_TIME.parse(text));
-		}
-		else {
-			// assuming UTC instant a la "2007-12-03T10:15:30.00Z"
-			return Instant.parse(text);
+		try {
+			return Instant.ofEpochMilli(Long.parseLong(text));
+		} catch (NumberFormatException e) {
+			if (text.length() > 0 && Character.isAlphabetic(text.charAt(0))) {
+				// assuming RFC-1123 value a la "Tue, 3 Jun 2008 11:05:30 GMT"
+				return Instant.from(DateTimeFormatter.RFC_1123_DATE_TIME.parse(text));
+			} else {
+				// assuming UTC instant a la "2007-12-03T10:15:30.00Z"
+				return Instant.parse(text);
+			}
 		}
 	}
 

--- a/spring-context/src/main/java/org/springframework/format/datetime/standard/InstantFormatter.java
+++ b/spring-context/src/main/java/org/springframework/format/datetime/standard/InstantFormatter.java
@@ -43,11 +43,13 @@ public class InstantFormatter implements Formatter<Instant> {
 	public Instant parse(String text, Locale locale) throws ParseException {
 		try {
 			return Instant.ofEpochMilli(Long.parseLong(text));
-		} catch (NumberFormatException e) {
+		}
+		catch (NumberFormatException ex) {
 			if (text.length() > 0 && Character.isAlphabetic(text.charAt(0))) {
 				// assuming RFC-1123 value a la "Tue, 3 Jun 2008 11:05:30 GMT"
 				return Instant.from(DateTimeFormatter.RFC_1123_DATE_TIME.parse(text));
-			} else {
+			}
+			else {
 				// assuming UTC instant a la "2007-12-03T10:15:30.00Z"
 				return Instant.parse(text);
 			}

--- a/spring-context/src/test/java/org/springframework/format/datetime/standard/DateTimeFormattingTests.java
+++ b/spring-context/src/test/java/org/springframework/format/datetime/standard/DateTimeFormattingTests.java
@@ -621,6 +621,19 @@ class DateTimeFormattingTests {
 								.hasMessageStartingWith("Text '210302'")
 								.hasNoCause();
 		}
+
+		@Test
+		void testBindInstantAsLongEpochMillis() {
+			MutablePropertyValues propertyValues = new MutablePropertyValues();
+			propertyValues.add("instant", 1234L);
+			binder.bind(propertyValues);
+			assertThat(binder.getBindingResult().getErrorCount()).isZero();
+			assertThat(binder.getBindingResult().getRawFieldValue("instant"))
+					.isInstanceOf(Instant.class)
+					.isEqualTo(Instant.ofEpochMilli(1234L));
+			assertThat(binder.getBindingResult().getFieldValue("instant"))
+					.hasToString("1970-01-01T00:00:01.234Z");
+		}
 	}
 
 

--- a/spring-context/src/test/java/org/springframework/format/datetime/standard/InstantFormatterTests.java
+++ b/spring-context/src/test/java/org/springframework/format/datetime/standard/InstantFormatterTests.java
@@ -140,7 +140,7 @@ class InstantFormatterTests {
 
 		@Override
 		public Stream<Arguments> provideArguments(ExtensionContext context) {
-			return random.longs(DATA_SET_SIZE, Instant.ofEpochMilli(Long.MIN_VALUE).toEpochMilli(), Instant.ofEpochMilli(Long.MAX_VALUE).toEpochMilli())
+			return random.longs(DATA_SET_SIZE, Long.MIN_VALUE, Long.MAX_VALUE)
 					.mapToObj(Instant::ofEpochMilli)
 					.map(instant -> instant.truncatedTo(ChronoUnit.MILLIS))
 					.map(Arguments::of);

--- a/spring-context/src/test/java/org/springframework/format/datetime/standard/InstantFormatterTests.java
+++ b/spring-context/src/test/java/org/springframework/format/datetime/standard/InstantFormatterTests.java
@@ -19,6 +19,7 @@ package org.springframework.format.datetime.standard;
 import java.text.ParseException;
 import java.time.Instant;
 import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
 import java.util.Random;
 import java.util.stream.Stream;
 
@@ -79,6 +80,16 @@ class InstantFormatterTests {
 		assertThat(actual).isEqualTo(expected);
 	}
 
+	@ParameterizedTest
+	@ArgumentsSource(RandomEpochMillisProvider.class)
+	void should_parse_into_an_Instant_from_epoch_mili(Instant input) throws ParseException {
+		Instant expected = input;
+
+		Instant actual = instantFormatter.parse(Long.toString(input.toEpochMilli()), null);
+
+		assertThat(actual).isEqualTo(expected);
+	}
+
 	private static class RandomInstantProvider implements ArgumentsProvider {
 
 		private static final long DATA_SET_SIZE = 10;
@@ -119,6 +130,20 @@ class InstantFormatterTests {
 		Stream<?> provideArguments() {
 			return randomInstantStream(min, max)
 					.map(DateTimeFormatter.RFC_1123_DATE_TIME.withZone(systemDefault())::format);
+		}
+	}
+	private static final class RandomEpochMillisProvider implements ArgumentsProvider {
+
+		private static final long DATA_SET_SIZE = 10;
+
+		private static final Random random = new Random();
+
+		@Override
+		public Stream<Arguments> provideArguments(ExtensionContext context) {
+			return random.longs(DATA_SET_SIZE, Instant.ofEpochMilli(Long.MIN_VALUE).toEpochMilli(), Instant.ofEpochMilli(Long.MAX_VALUE).toEpochMilli())
+					.mapToObj(Instant::ofEpochMilli)
+					.map(instant -> instant.truncatedTo(ChronoUnit.MILLIS))
+					.map(Arguments::of);
 		}
 	}
 

